compiler/varpartitions.nim

```diff
@@ -457,11 +457,11 @@ proc destMightOwn(c: var Partitions; dest: var VarIndex; n: PNode) =
     destMightOwn(c, dest, n[0])
 
   of nkCallKinds:
-    if hasDestructor(n.typ):
```

Comment:
There is no need to check nil twice IMO.

```nim
  if n.typ != nil:
    if hasDestructor(n.typ):
      # calls do construct, what we construct must be destroyed,
      # so dest cannot be a cursor:
      dest.flags.incl ownsData
    elif n.typ.kind in {tyLent, tyVar}:
      # we know the result is derived from the first argument:
      var roots: seq[PSym]
      allRoots(n[1], roots)
      for r in roots:
        connect(c, dest.sym, r, n[1].info)
```

Reply:
Then you need additional logic to make sure that both `nil` types and types that don't match either condition reach the following `else`. This may actually not matter for `noreturn` procs, but I'm not certain, so I left that bit unchanged.